bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java

```diff
@@ -960,19 +974,31 @@ public void run() {
                             }
                             journalFlushWatcher.reset().start();
                             bc.flush(false);
+
+                            for (int i = 0; i < toFlush.size(); i++) {
```

Comment:
any reason why you still keep the for-loop here? There should be no code added between line 975 and line 988, it is registering the time journal thread spends on flushing.

I suggested you merge this for-loop with line 1005-1038, to keep branches handling close to each other. any reason why you didn't do that?

Reply:
This is because I have to call the callback "before" line 1007, that is requesting the fsync the file and firing the callbacks.

```
forceWriteRequests.put(createForceWriteRequest(logFile, logId, lastFlushPosition,
                                                                               toFlush, shouldRolloverJournal, false));
```

If I fire the callbacks "after" that line the callback could be fired by the ForceWriteThread as well.

So I enqueue the callbacks (QueueEntry) into the threadpool and remove them from the toFlush list.
The ForceWriteThread will find nulls for callbacks already fired